### PR TITLE
Feature/servers

### PR DESCRIPTION
**Pull request by wayland**
- In configuration.sql:
  - Added a generic `configuration` table -- this can probably be removed once we have a configuration file
  - Added a `servers` table for configuring the Servers block.  The table contains a slug and a priority.  The Servers block will contain one item for each slug, and will choose the one with the highest priority (see tests for examples).  
  - Added `openapi_server_objects` which returns JSON based on the servers table
  - Added `set_server_from_configuration`, which updates the `servers` table based on the configuration values
- In main.sql:
  - Added `callable_root` function which can be directly called by PostgREST
  - Added `x-software` block, for specifying software versions
- In fixtures.sql:
  - For testing, added a pre_config function that sets the `pgrst.server_host` and `pgrst.server_port` (and then called it)
- Tests:
  - Added some tests for `servers` and `main`

### CODE REVIEW

**Comment by wayland** [comment]
Hi all!  Here's my PR for the servers/x-software stuff.  I'm expecting to be told I've done a number of things wrong.  I probably won't have a chance to look at the corrections for another week.  

Thanks!

**Comment by wayland** [comment]
Interesting!  The tests are failing.  They worked on my local (I'm testing with a pretty recent Postgres, though).  I note the following error which wasn't present on my local.  
```
The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
```
I'm not familiar enough with nix to figure out what's going on.  

Question: is anyone else able to replicate the test failures on their local?

**Comment by steve-chavez** [comment]
@wayland Yes, I'm able to replicate the errors locally. (I do realize the CI is not being informative, will fix)

The error happens on pg 14 and 13, the output varies there:

```diff
diff -U3 /home/stevechavez/Projects/postgrest-openapi/test/expected/servers.out /home/stevechavez/Projects/postgrest-openapi/output/results/servers.out
--- /home/stevechavez/Projects/postgrest-openapi/test/expected/servers.out	2023-11-04 18:19:12.619939571 -0500
+++ /home/stevechavez/Projects/postgrest-openapi/output/results/servers.out	2023-11-04 18:21:30.704413555 -0500
@@ -53,16 +53,16 @@
 -----------------------------------------------------------------------------
  [                                                                          +
      {                                                                      +
-         "url": "https://www.example.com/api/",                             +
-         "description": "Overriding URL"                                    +
-     },                                                                     +
-     {                                                                      +
          "url": "https://www.example.com/otherapi/?s={s}&browser={browser}",+
          "variables": {                                                     +
              "s": "worms",                                                  +
              "browser": "none"                                              +
          },                                                                 +
          "description": "Additional URL"                                    +
+     },                                                                     +
+     {                                                                      +
+         "url": "https://www.example.com/api/",                             +
+         "description": "Overriding URL"                                    +
      }                                                                      +
  ]
 (1 row)
```

File: test/expected/servers.out [hunk 1]
```diff
@@ -1,22 +1,79 @@
...
+         "description": "Overriding URL"                                    +
+     },                                                                     +
+     {                                                                      +
+         "url": "https://www.example.com/otherapi/?s={s}&browser={browser}",+
+         "variables": {                                                     +
+             "s": "worms",                                                  +
+             "browser": "none"                                              +
+         },                                                                 +
+         "description": "Additional URL"                                    +
+     }                                                                      +
```
Comment (steve-chavez) [inline test/expected/servers.out:66]: The error is here. The order of the json objects varies between pg versions. An explicit ORDER BY seems like the solution.

**Comment by steve-chavez** [comment]
@wayland Just merged https://github.com/PostgREST/postgrest-openapi/pull/23. Once you rebase you should be able to see the diff with the failures.

**Comment by wayland** [comment]
Thanks!  I've added the ORDER BY, so it should work now.  Looks like the tests pass now.  

HTH,

File: sql/servers.sql [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+	priority	INTEGER NOT NULL,
+	PRIMARY KEY (slug, priority)
+);
+INSERT INTO servers (url, description, priority) VALUES ('http://0.0.0.0:3000/', 'Default URL', 10);
+/*
+COPY servers (url, description, priority) FROM stdin;
+http://0.0.0.0:80/	Default URL	10
+http://0.0.0.0:80/	Default URL	20
+\.
+*/
```
Comment (laurenceisla) [inline sql/servers.sql:68]: Ditto

```suggestion
```

File: sql/main.sql [hunk 3]
```diff
@@ -1,9 +1,23 @@
 -- Default PostgREST OpenAPI Specification
 
+-- This is the one we should be calling
+create or replace function callable_root() returns jsonb as $$
+	-- Calling this every time is inefficient, but it's the best we can do until PostgREST calls it when it updates the server config
+	CALL set_server_from_configuration();
+	SELECT get_postgrest_openapi_spec(
+		schemas := string_to_array(current_setting('pgrst.db_schemas', TRUE), ','),
+		postgrest_version := current_setting('pgrst.version', TRUE),
+		proa_version := '0.1'::text, -- FIX: needs to be updated; put into config, and have Makefile update
```
Comment (laurenceisla) [inline sql/main.sql:10]: I think these `FIX`es should be preceded by `TODO`, most IDEs highlight/summarize the TODOs nicely. Something like:

```suggestion
		proa_version := '0.1'::text, -- TODO: (FIX) needs to be updated; put into config, and have Makefile update
```

File: test/sql/main.sql [hunk 4]
```diff
@@ -0,0 +1,24 @@
+-- Test get_postgrest_openapi_spec
+--select jsonb_pretty(get_postgrest_openapi_spec('{test}')->'servers');
+
```
Comment (laurenceisla) [inline test/sql/main.sql:3]:
```suggestion
```

**Review by laurenceisla — COMMENTED** [review]
Great job on the feature. I think we should establish some styling guides for the project in the structure, comments, code, etc. along the way. For instance:

File: sql/main.sql [hunk 5]
```diff
@@ -1,9 +1,23 @@
 -- Default PostgREST OpenAPI Specification
 
+-- This is the one we should be calling
+create or replace function callable_root() returns jsonb as $$
+	-- Calling this every time is inefficient, but it's the best we can do until PostgREST calls it when it updates the server config
+	CALL set_server_from_configuration();
+	SELECT get_postgrest_openapi_spec(
+		schemas := string_to_array(current_setting('pgrst.db_schemas', TRUE), ','),
```
[the remark this answers is not in the document]
Comment (wayland) [reply sql/main.sql]: Done

File: sql/servers.sql [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+INSERT INTO configuration (name, value) VALUES ('api-version', '');
+INSERT INTO configuration (name, value) VALUES ('api-description', '');
+/*
+COPY configuration (name, value) FROM STDIN;
+mode	follow-privileges
+security-active	false
+api-name	
+api-description	
+\.
+*/
```
[the remark this answers is not in the document]
Comment (wayland) [reply sql/servers.sql]: Removed.

File: sql/servers.sql (hunk 2 above)
Comment (wayland) [reply sql/servers.sql]: Removed

File: sql/main.sql (hunk 3 above)
Comment (wayland) [reply sql/main.sql]: Did s/FIX/TODO/ and s/FIXMVP/TODO: (MVP)/

File: test/sql/main.sql (hunk 4 above)
Comment (wayland) [reply test/sql/main.sql]: Removed.

**Comment by wayland** [comment]
Thanks for your feedback.  I've done some edits in line with what you've suggested.  Let me know if there are further issues.

**Review by laurenceisla — APPROVED** [review]
Nice! LGTM!

**Review by steve-chavez — APPROVED** [review]

**Pull request merged**